lib/codecs/src/decoding/format/vrl.rs

```diff
@@ -0,0 +1,257 @@
+use crate::decoding::format::Deserializer;
+use crate::BytesDeserializerConfig;
+use bytes::Bytes;
+use derivative::Derivative;
+use smallvec::{smallvec, SmallVec};
+use vector_config_macros::configurable_component;
+use vector_core::config::{DataType, LogNamespace};
+use vector_core::event::{Event, TargetEvents, VrlTarget};
+use vector_core::{compile_vrl, schema};
+use vrl::compiler::state::ExternalEnv;
+use vrl::compiler::{runtime::Runtime, CompileConfig, Program, TimeZone, TypeState};
+use vrl::diagnostic::Formatter;
+use vrl::value::Kind;
+
+/// Config used to build a `VrlDeserializer`.
+#[configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct VrlDeserializerConfig {
+    /// VRL-specific decoding options.
+    vrl: VrlDeserializerOptions,
+}
+
+/// VRL-specific decoding options.
+#[configurable_component]
+#[derive(Debug, Clone, PartialEq, Eq, Derivative)]
+#[derivative(Default)]
+pub struct VrlDeserializerOptions {
+    /// The [Vector Remap Language][vrl] (VRL) program to execute for each event.
+    ///
+    /// [vrl]: https://vector.dev/docs/reference/vrl
+    pub source: String,
+
+    /// The name of the timezone to apply to timestamp conversions that do not contain an explicit
+    /// time zone. The time zone name may be any name in the [TZ database][tz_database], or `local`
+    /// to indicate system local time.
+    ///
+    /// If not set, `local` will be used.
+    ///
+    /// [tz_database]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
+    #[serde(default)]
+    #[configurable(metadata(docs::advanced))]
+    pub timezone: Option<TimeZone>,
+}
+
+impl VrlDeserializerConfig {
+    /// Build the `VrlDeserializer` from this configuration.
+    pub fn build(&self) -> vector_common::Result<VrlDeserializer> {
+        let state = TypeState {
+            local: Default::default(),
+            external: ExternalEnv::default(),
+        };
+
+        match compile_vrl(
+            &self.vrl.source,
+            &vrl::stdlib::all(),
+            &state,
+            CompileConfig::default(),
+        ) {
+            Ok(result) => Ok(VrlDeserializer {
+                program: result.program,
+                timezone: self.vrl.timezone.unwrap_or(TimeZone::Local),
+            }),
+            Err(diagnostics) => Err(Formatter::new(&self.vrl.source, diagnostics)
+                .to_string()
+                .into()),
+        }
```

Comment:
nit: This could be done with `map` and `map_err` if you were so inclined. It doesn't look like it's any simpler, so TIOLI.
```suggestion
        compile_vrl(
            &self.vrl.source,
            &vrl::stdlib::all(),
            &state,
            CompileConfig::default(),
        )
        .map(|result|VrlDeserializer {
            program: result.program,
            timezone: self.vrl.timezone.unwrap_or(TimeZone::Local),
        })
        .map_err(|diagnostics| Formatter::new(&self.vrl.source, diagnostics)
            .to_string()
            .into()),
        )
```